src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsAdornmentManager.cs

```diff
@@ -0,0 +1,237 @@
+                    var graphicsResult = tag.GetGraphics(TextView, geometry, GetFormat(classificationType));
+                    if (!tagSpanToPointMap.TryGetValue(tagMappingSpan, out var point))
+                    {
+                        continue;
+                    }
+
+                    var lineView = TextView.GetTextViewLineContainingBufferPosition(point);
+
+                    var visualElement = graphicsResult.VisualElement;
+                    if (tag.Location is InlineDiagnosticsLocations.PlacedAtEndOfCode)
+                    {
+                        Canvas.SetLeft(visualElement, lineView.Right);
+                    }
+                    else if (tag.Location is InlineDiagnosticsLocations.PlacedAtEndOfEditor)
+                    {
+                        Canvas.SetLeft(visualElement, TextView.ViewportWidth - visualElement.DesiredSize.Width);
+                    }
+
+                    tag.EditorTextLocation = lineView.Right;
+                    Canvas.SetTop(visualElement, geometry.Bounds.Bottom - visualElement.DesiredSize.Height);
+
+                    if (lineView.Right < TextView.ViewportWidth - visualElement.DesiredSize.Width)
+                    {
+                        AdornmentLayer.AddAdornment(
+                            // TextRelative because the adornment is positioned relative to the text in the editor
+                            behavior: AdornmentPositioningBehavior.TextRelative,
+                            visualSpan: span,
+                            tag: tag,
+                            adornment: visualElement,
+                            removedCallback: delegate { graphicsResult.Dispose(); });
+                    }
+                }
+            }
+        }
+
+        private static Dictionary<IMappingTagSpan<InlineDiagnosticsTag>, SnapshotPoint> GetTagSpansToSnapshotPointMap(
+            IDictionary<(int linenum, SnapshotSpan changedSpan), List<IMappingTagSpan<InlineDiagnosticsTag>>> map)
+        {
+            var tagSpanToPointMap = new Dictionary<IMappingTagSpan<InlineDiagnosticsTag>, SnapshotPoint>();
+
+            foreach (var kvp in map)
+            {
+                foreach (var mappingTagSpan in kvp.Value)
+                {
+                    var point = mappingTagSpan.Span.Start.GetPoint(kvp.Key.changedSpan.Snapshot, PositionAffinity.Predecessor);
+                    if (point == null)
+                    {
+                        continue;
+                    }
+
+                    tagSpanToPointMap.Add(mappingTagSpan, point.Value);
+                }
+            }
+
+            return tagSpanToPointMap;
+        }
+
+        private static IMappingTagSpan<InlineDiagnosticsTag> GetHighestOrderTag(List<IMappingTagSpan<InlineDiagnosticsTag>> list)
+        {
+            return list.Where(s => s.Tag.ErrorType is PredefinedErrorTypeNames.SyntaxError).FirstOrDefault() ?? list.First();
```

Comment:
FirstOrDefault should take a predciate right?

Reply:
removed method entirely